src/bootstrap/src/core/builder/mod.rs

```diff
@@ -131,7 +135,26 @@ pub trait Step: 'static + Clone + Debug + PartialEq + Eq + Hash {
     /// depending on the `Step::run` implementation of the caller.
     fn run(self, builder: &Builder<'_>) -> Self::Output;
 
-    /// Determines if this `Step` should be run when given specific paths (e.g., `x build $path`).
+    /// Some steps that are marked `DEFAULT = true` want to only be treated as
+    /// default steps on a conditional basis, depending on factors like config
+    /// settings or the availability of external tools.
+    ///
+    /// Those steps can override this function, which will be called just before
+    /// the step would actually be run as part of the default set of steps, so
+    /// that the step can conditionally decide whether to run or not.
+    ///
+    /// If the underlying check should not be performed repeatedly
+    /// (e.g. because it probes command-line tools),
+    /// consider memoizing its outcome via a field in the builder.
+    fn is_really_default(builder: &Builder<'_>) -> bool {
+        // This default impl doesn't use the builder, but we want it to be
+        // named `builder` for IDE autocompletion.
+        let _ = builder;
+        Self::DEFAULT
+    }
```

Comment:
Discussion: maybe a name like `finalized_default`? I kinda hate that name too, it reminds me of the convoluted compiler lint level decoration logic... Anyway, 'finalized' in the sense that there's a hierarchy or precedence ordering:

1. `Step::DEFAULT` is the "base" default.
2. `is_really_default` or `finalized_default` can "override" this base default subject to conditions.

(2) then finally determines if a `Step` is eligible to run.

I say this because `is_really_default` I find quite confusing to read...

Reply:
Would it be possible to merge `const DEFAULT` and `fn is_really_default` into a single `fn default` that returns a constant value for most steps?

Reply:
Hm yeah. I think I like that better.